src/allmydata/test/test_testing.py

```diff
@@ -0,0 +1,100 @@
+# -*- coding: utf-8 -*-
+# Tahoe-LAFS -- secure, distributed storage grid
+#
+# Copyright © 2020 The Tahoe-LAFS Software Foundation
+#
+# This file is part of Tahoe-LAFS.
+#
+# See the docs/about.rst file for licensing information.
+
+"""
+Tests for the allmydata.testing helpers
+"""
+
+from twisted.internet.defer import (
+    inlineCallbacks,
+)
+
+from allmydata.uri import (
+    from_string,
+    CHKFileURI,
+)
+from allmydata.testing.web import (
+    create_tahoe_treq_client,
+)
+
+from hypothesis import (
+    given,
+)
+from hypothesis.strategies import (
+    binary,
+)
+
+from testtools import (
+    TestCase,
+)
+from testtools.matchers import (
+    Always,
+)
+from testtools.twistedsupport import (
+    succeeded,
+)
+
+
+class FakeWebTest(TestCase):
+    """
+    Test the WebUI verified-fakes infrastucture
+    """
+
+    def setUp(self):
+        super(FakeWebTest, self).setUp()
+        self.http_client = create_tahoe_treq_client()
+
+    @given(
+        content=binary(),
+    )
+    def test_create_and_download(self, content):
+        """
+        Upload some content (via 'PUT /uri') and then download it (via
+        'GET /uri?uri=...')
+        """
+
+        @inlineCallbacks
+        def do_test():
+            resp = yield self.http_client.put("http://example.com/uri?replace=true", content)
```

Comment:
I went through the `frontend/webapi.`rst docs and looked at the "real" implementation to try to understand this `replace=true` thing better (the very idea of a "replace" operation struck me as quite surprising considering my assumptions about Tahoe-LAFS and the fact that keys are cryptographically derived from content ... how could replacement be meaningful at all in that case...).

What I found is that the docs are as inscrutable as ever, though perhaps as good as can reasonably be expected considering how complex the interface is.

From reading the implementation, I wasn't able to determine that the `replace` query argument has any meaning *at all* to the `/uri` handler, though.  Instead, it seems to be meaningful for `DirectoryNodeHandler` where it means a child with the same name may be replaced.  I think this inconsistency is a hint that the fake is slightly too simple to emulate the behaviors it is trying to emulate.  So, specifically, against the real implementation, `PUT /uri?replace=true` seems like it will behave exactly as `PUT /uri` and `PUT /uri?replace=false`.  I hope I haven't totally misread the implementation...

Maybe this PR shouldn't try to support as much of the /uri hierarchy as it is?  If it supported only PUT to /uri exactly then there is no `replace` query argument (the only args are `uri` for GET and `t` for PUT and POST, I think - eventually it would be good to enforce this in the fake).

Reply:
Okay, I will remove all the `replace=` code.

It was added because the previous review wanted more error-cases handled (...and I think only because I had too much "defensive coding" in the `add_data` method which checked if a key was already there, but that should never happen since we make a new fake cap every time). That is: https://github.com/tahoe-lafs/tahoe-lafs/pull/728#discussion_r439584285 and https://github.com/tahoe-lafs/tahoe-lafs/pull/728#discussion_r439584374

So, armed with your new understanding of the Tahoe code ;) we simply can never hit any "conflict" cases, right? Because PUT doesn't error and we're only ever creating immutable-directories (so we never "add" any new children to anything).